// testcafe/lib/proxyless/client/event-simulator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../types");
// @ts-ignore
const hammerhead_1 = require("../../client/core/deps/hammerhead");
const utils_1 = require("./utils");
const axis_values_1 = __importDefault(require("../../client/core/utils/values/axis-values"));
const MOUSE_EVENT_OPTIONS = {
    clickCount: 1,
    button: 'left',
};
class ProxylessEventSimulator {
    constructor(dispatchEventFn, leftTopPoint) {
        this._dispatchEventFn = dispatchEventFn;
        this._leftTopPoint = leftTopPoint || new axis_values_1.default(0, 0);
    }
    _createMouseEventOptions(type, options) {
        return hammerhead_1.utils.extend({
            x: options.options.clientX + this._leftTopPoint.x,
            y: options.options.clientY + this._leftTopPoint.y,
            modifiers: (0, utils_1.calculateKeyModifiersValue)(options.options),
            button: (0, utils_1.calculateMouseButtonValue)(options.options),
            type,
        }, MOUSE_EVENT_OPTIONS);
    }
    mouseDown(options) {
        const eventOptions = this._createMouseEventOptions('mousePressed', options);
        return this._dispatchEventFn(types_1.EventType.Mouse, eventOptions);
    }
    mouseUp(options) {
        const eventOptions = this._createMouseEventOptions('mouseReleased', options);
        return this._dispatchEventFn(types_1.EventType.Mouse, eventOptions);
    }
}
exports.default = ProxylessEventSimulator;
module.exports = exports.default;
